Implement sleep failpoint status/counter for linearizability test cases

ramil600 opened the pull request:
Each time sleep is executed in actSleep counter is incremented. Added counter endpoint that gives number of sleeps executed on the current terms: GET pkg/failpoint/count When DELETE /pkg/failpoint is executed, failpoint.terms is nilled, hence counter is reset.

Requested in reference to:
tests/linearizability: added sleep fail point injection #14796

Signed-off-by: Ramil Mirhasanov <ramil600@yahoo.com>

ramil600 commented:
cc @serathius

serathius commented on runtime/http.go:
```diff
@@ -72,6 +72,13 @@ func (*httpHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			}
 			w.Write([]byte(strings.Join(lines, "\n") + "\n"))
+		} else if strings.HasSuffix(key, "/count") {
+			fp := key[:len(key)-6]
```
```suggestion
			fp := key[:len(key)-len("/count")]
```

serathius commented on runtime/runtime.go:
```diff
@@ -114,6 +114,25 @@ func Status(failpath string) (string, error) {
 	return t.desc, nil
 }
 
+// StatusCount gives the number of sleeps executed on the current terms
```
```suggestion
// StatusCount gives the how many times current term was executed
```

serathius commented on runtime/terms.go:
```diff
@@ -41,6 +41,8 @@ type terms struct {
 
 	// mu protects the state of the terms chain
 	mu sync.Mutex
+	// counts amount of sleep executions
```
```suggestion
	// counts amount of executions
```

serathius commented on runtime/terms.go:
```diff
@@ -310,6 +312,9 @@ func actSleep(t *term) interface{} {
 	}
 	time.Sleep(dur)
+	t.parent.mu.Lock()
+	t.parent.counter++
```
Could we count number of any executions instead of just sleep?

serathius commented:
Overall looks good, however let's count any execution instead of only sleep ones.

ramil600 commented:
@serathius, implemented fixes as per your suggestions. 
I added a test that follows terms and executes failpoints accordingly, the only problem is that it is failing linter check, because I am using gofail syntax. Please advise how should I go about that. Do I need to build a script like for linearizability tests?

serathius commented on runtime/http.go:
```diff
@@ -72,6 +72,13 @@ func (*httpHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 				lines[i] = fps[i] + "=" + s
 			}
 			w.Write([]byte(strings.Join(lines, "\n") + "\n"))
+		} else if strings.HasSuffix(key, "/count") {
```
API could be improved, but this should work for now.

serathius commented on runtime/runtime.go:
```diff
@@ -114,6 +114,25 @@ func Status(failpath string) (string, error) {
 }
 
+// StatusCount outputs how many times current term was executed
+func StatusCount(failpath string) (string, error) {
```
I would expect that `StatusCount` function should return an `int` instead of `string`. However that can be fixed in future PRs.

serathius commented on runtime/termscounter_test.go:
```diff
@@ -0,0 +1,60 @@
+			name: "runtime_test/ExampleString",
+			desc: `10*sleep(10)->1*return("abc")`,
+			run:  12,
+			want: "11",
```
If you run loop `12` times, how the result is `11`?

ramil600 replied:
this happens because we are building terms here by saying: 
`desc: `10*sleep(10)->1*return("abc")`,`
sleep exactly 10 times and then return 1 time.
Even if the function is called more than 11 times, no terms will be executed, because its controlled by internal counter: 
`func (mc *modCount) allow() bool {
	if mc.c > 0 {
		mc.c--
		return true
	}
	return false
}` 
That way we can check if the fail point is executed exactly 11 times and not the number of function calls.

serathius commented on runtime/termscounter_test.go:
```diff
@@ -0,0 +1,60 @@
+		want string
+	}{
+		{
+			name: "runtime_test/ExampleString",
```
Could you add more tests scenarios? For example: Before enable, after disable, after changing term etc.

ramil600 replied:
done

serathius commented on runtime/termscounter_test.go:
```diff
@@ -0,0 +1,60 @@
+		},
+	}
+
+	for _, tc := range testcases {
```
If you want to use table tests, please use subtests.

serathius commented:
In this case I think it's ok to disable the lint rule on those lines. For example
```
var __fp_ExampleString *runtime.Failpoint = runtime.NewFailpoint("runtime_test", "ExampleString") //nolint:stylecheck
```

ahrtr commented on runtime/terms.go:
```diff
@@ -41,6 +41,8 @@ type terms struct {
 	// mu protects the state of the terms chain
 	mu sync.Mutex
+	// counts executions
+	counter int
```
Why add this field?

ahrtr commented:
Note: when you implement any new feature or do any enhancement, please clarify the requirement first. I do not get the info from the description. 

Please also make sure you read [design.md](https://github.com/etcd-io/gofail/blob/master/doc/design.md). Note that the existing design & implementation already supports counter. For example,
```
2*sleep("10s")  # It means the sleep will be executed twice
2*return("abc")  # it means the return will be executed twice
```

serathius commented:
> Note: when you implement any new feature or do any enhancement, please clarify the requirement first. I do not get the info from the description.

Requirements were discussed in https://github.com/etcd-io/etcd/pull/14796, unfortunatelly issue was not correctly linked.

ahrtr commented:
If you really need more visibility of the terms, I would suggest to enhance the existing `Status` function. Currently it only [returns the description](https://github.com/etcd-io/gofail/blob/51e235eacea6c6b1744eb04e40f19575b54758f0/runtime/runtime.go#L114). We can define a struct or format, and get more info included.

lavacat commented:
@ramil600 are you still working on this PR? No rush, but please reply, otherwise I'll take over in couple days.

ramil600 commented:
@lavacat, sorry for late reply, I was working on it, but there was no resolution from maintainers, on whether my PR should be merged. If you can help me, I am willing complete this one, please review my submitted PR.

lavacat commented:
@ramil600 I've looked at the PR.
1. Per @ahrtr suggestion we should use runtime.Status
2. Maybe we should just count evals and not actual term executions. Different terms in term chain can be executed different number of times. Seams confusing. But I'm open to keeping it as is.
3. Please update PR description, your original one is about `sleep`. I think it can be worded better.
4. There are some comments from @serathius and @ahrtr that aren't resolved. Please ask to clarify or resolve.

ahrtr commented:
> 1. Per @ahrtr suggestion we should use runtime.Status

Yes, please feel free to follow this direction to work on this task. But please keep it backward compatible, which means it shouldn't break anything on the existing etcd's linearizablity test.



> 2\. Different terms in term chain can be executed different number of times. Seams confusing. But I'm open to keeping it as is.

The existing logic on term isn't perfect, but it meets the requirement of the etcd test for now. We can enhance it on demand. I thought about refactoring& enhancing the overall design (e.g. adding `AND`, `OR` logic to control the execution of term chain), but I gave up because it is neither a priority for now, nor we receive such request so far. Anyway, we can revisit this when there is a real use case.

ramil600 commented:
Thank you guys, @ahrtr, @lavacat,  if this is not high priority, I will look into it and get back within few days.

ramil600 commented:
> If you really need more visibility of the terms, I would suggest to enhance the existing `Status` function. Currently it only [returns the description](https://github.com/etcd-io/gofail/blob/51e235eacea6c6b1744eb04e40f19575b54758f0/runtime/runtime.go#L114). We can define a struct or format, and get more info included.

In order to keep the function signature same
Could we return remaining terms description for sleep fail point, instead of description?
Let's say if sleep failpoint was already executed five times out of ten, we could return: 
5*sleep(10)->1*return("abc")
if counter is reset after all the executions we can also leave it but mark it as 0:
0*sleep(10)->1*return("abc").

@ahrtr, please let me know what do you think about it?

Currently I added StatusCount() function in the same PR, that counts executions with help of new terms.counter variable.
That is alternative implementation.

lavacat commented:
> Could we return remaining terms description for sleep fail point, instead of description?

I think it will break logic for checking if failpoint is [Available](https://github.com/etcd-io/etcd/blob/main/tests/robustness/failpoints.go#L287)
Also, can be confusing for new users.

IMO, we should define Status output as json and then refactor etcd [fetchFailpoints](https://github.com/etcd-io/etcd/blob/main/tests/framework/e2e/etcd_process.go#LL335C6-L335C21)

Since this is blocking sleep failpoints impl I'm ok with merging current `/count`. We can discuss format and do the refactor later. Etcd community owns both repos and gofail used mostly by etcd, so probably we can drop `/count` without backward compatibility. CC @ahrtr @serathius 

@ramil600 please also address

lavacat commented:
@ramil600 are you still working on this? If you won't reply by May 15th, I'll assume you are busy  and I'll look for somebody to take over. You'll still be the author of the commit. IMO this is very close to done if we stick with `/count`.

pchan commented:
> @ramil600 are you still working on this? If you won't reply by May 15th, I'll assume you are busy and I'll look for somebody to take over. You'll still be the author of the commit. IMO this is very close to done if we stick with `/count`.

I have attempted to update Ramil's code to run successfully in https://github.com/etcd-io/gofail/pull/47. If Ramil wants to take it up he can continue or I can start fixing issues listed in https://github.com/etcd-io/gofail/pull/37#issuecomment-1380866250. @ramil600 is still the author of the original commit.

serathius commented:
Closing for https://github.com/etcd-io/gofail/pull/47
